lang/java/avro/src/main/java/org/apache/avro/reflect/ReflectDatumReader.java

```diff
@@ -123,22 +116,37 @@ protected void addToArray(Object array, long pos, Object e) {
    * representations.
    */
   protected Object readArray(Object old, Schema expected, ResolvingDecoder in) throws IOException {
-    Schema expectedType = expected.getElementType();
+    return readArray(old, null, expected, in);
+  }
+
+  private Object readArray(Object old, Type type, Schema expected, ResolvingDecoder in) throws IOException {
+    Class<?> elementClass = null;
+    if (type instanceof Class && ((Class<?>) type).isArray()) {
+      // get array element class
+      elementClass = ((Class<?>) type).getComponentType();
+    } else if (type instanceof ParameterizedType
+        && Collection.class.isAssignableFrom((Class<?>) ((ParameterizedType) type).getRawType())) {
+      // get collection element class
+      elementClass = (Class<?>) ((ParameterizedType) type).getActualTypeArguments()[0];
+    }
```

Comment:
Nicely done!

This logic also occurs in `readAndConvert` as well, and I think it would be easier to understand if extracted info a method.

Reply:
the logic is little different, I've extracted them into two private methods `getConversion` and `getElementClass`. The logic here is to get element class of arrays and collections, the logic in readAndConvert is to get the raw type of parametrized classes and unparametrized classes (including arrays).